### PR TITLE
[SQL] Adding AAD-only Support for SQL Managed Instances and Servers

### PR DESCRIPTION
**Pull request by jaredwelsh**
Adding support for a new feature to az cli. Azure Active Directory only allows for only AD users to be admins not traditional username/password combinations. This feature has been added to Azure Python SDK and is now being supported in az cli

Using either a managed instance or server with a ad admin run:

az sql server ad-only-auth enable
az sql server ad-only-auth disable
az sql server ad-only-auth get

az sql mi ad-only-auth enable
az sql mi ad-only-auth disable
az sql mi ad-only-auth get

to enable, disable and show AAD only authentication.

### CODE REVIEW

**Comment by strehan1993** [comment]
Looks good.

**Comment by yonzhan** [comment]
SQL

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 1]
```diff
@@ -630,6 +630,26 @@
...
+
+helps['sql mi ad-only-auth disable'] = """
+type: command
+short-summary: Disable Azure Active Directly only Authentication for this Managed Instance.
+"""
+
+helps['sql mi ad-only-auth get'] = """
+type: command
+short-summary: Get a specific Azure Active Directly only Authentication property.
+"""
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py:651]: Could you add example for the three commands?

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 2]
```diff
@@ -915,6 +935,31 @@
...
+
+helps['sql server ad-only-auth disable'] = """
+type: command
+short-summary: Disable Azure Active Directly only Authentication for this Server.
+"""
+
+helps['sql server ad-only-auth get'] = """
+type: command
+short-summary: Get a specific Azure Active Directly only Authentication property.
+"""
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py:961]: same as above

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 3]
```diff
@@ -570,6 +572,18 @@ def load_command_table(self, _):
...
+    server_aadonly_operations = CliCommandType(
+        operations_tmpl='azure.mgmt.sql.operations#ServerAzureADOnlyAuthenticationsOperations.{}',
+        client_factory=get_sql_server_azure_ad_only_operations)
+
+    with self.command_group('sql server ad-only-auth',
+                            server_aadonly_operations,
+                            client_factory=get_sql_server_azure_ad_only_operations) as g:
+
+        g.custom_command('disable', 'server_aad_only_disable')
+        g.custom_command('enable', 'server_aad_only_enable')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py:584]: Are they long running operations?
If yes, we will recommend you to add support_no_wait=True. See
https://github.com/Azure/azure-cli/blob/1ef59bace4dfad9cba137f49c8974aca4b0fff90/src/azure-cli/azure/cli/command_modules/sql/commands.py#L118

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 4]
```diff
@@ -570,6 +572,18 @@ def load_command_table(self, _):
...
+        g.command('get', 'get')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py:585]: CLI uses show command to do get operation. See  
https://github.com/Azure/azure-cli/blob/1ef59bace4dfad9cba137f49c8974aca4b0fff90/src/azure-cli/azure/cli/command_modules/sql/commands.py#L132

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 5]
```diff
@@ -636,6 +650,18 @@ def load_command_table(self, _):
...
+        operations_tmpl='azure.mgmt.sql.operations#ManagedInstanceAzureADOnlyAuthenticationsOperations.{}',
+        client_factory=get_sql_managed_instance_azure_ad_only_operations)
+
+    with self.command_group('sql mi ad-only-auth',
+                            managed_instance_aadonly_operations,
+                            client_factory=get_sql_managed_instance_azure_ad_only_operations) as g:
+
+        g.custom_command('disable', 'mi_aad_only_disable')
+        g.custom_command('enable', 'mi_aad_only_enable')
+        g.command('get', 'get')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py:663]: same as above

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 6]
```diff
@@ -2911,11 +2946,48 @@ def mi_ad_admin_delete(
...
+
+def mi_aad_only_disable(
+        client,
+        resource_group_name,
+        managed_instance_name):
+    '''
+    Disables the managed instance AAD-only setting
+    '''
+
+    return client.create_or_update(
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:2969]: If it is long running operation and you want to support no wait, besides adding support_no_wait_true, you also need to wrap the operation like
https://github.com/Azure/azure-cli/blob/1ef59bace4dfad9cba137f49c8974aca4b0fff90/src/azure-cli/azure/cli/command_modules/sql/custom.py#L813-L818

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 7]
```diff
@@ -2911,11 +2946,48 @@ def mi_ad_admin_delete(
...
+
+def mi_aad_only_enable(
+        client,
+        resource_group_name,
+        managed_instance_name):
+    '''
+    Enables the AAD-only setting
+    '''
+
+    return client.create_or_update(
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:2984]: same

**Review by Juliehzl — CHANGES_REQUESTED** [review]
please add test scenarios for all new commands and parameters.

**Review by Juliehzl — CHANGES_REQUESTED** [review]
Could you also add test scenarios for your new commands?

File: src/azure-cli/azure/cli/command_modules/sql/commands.py (hunk 3 above)
Comment (jaredwelsh) [reply src/azure-cli/azure/cli/command_modules/sql/commands.py:584]: they are not long running operations and therefore do not need to support no_wait

**Comment by jaredwelsh** [comment]
Sure. I'll write them up.

**Comment by jaredwelsh** [comment]
Tests were added

**Comment by jaredwelsh** [comment]
Style updates

**Comment by jaredwelsh** [comment]
@Juliehzl I don't think that error is caused by me. according to the logs the error comes from loading ai-examples which I haven't touched. Please give guidance on how to proceed.

File: src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 8]
```diff
@@ -971,6 +971,105 @@ def test_aad_admin(self, resource_group, server):
...
+
+        self.cmd('sql server ad-only-auth enable -n {} -g {}'.format(server, resource_group), checks=[])
+        self.cmd('sql server ad-only-auth disable -n {} -g {}'.format(server, resource_group), checks=[])
+        self.cmd('sql server ad-only-auth get -n {} -g {}'.format(server, resource_group), checks=[])
+
+
+class SqlManagedInstanceAzureActiveDirectoryAdministratorScenarioTest(ScenarioTest):
+
+    # Remove when issue #9393 is fixed.
+    @live_only()
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py:995]: Could we remove live_only() here?

**Comment by Juliehzl** [comment]
> @Juliehzl I don't think that error is caused by me. according to the logs the error comes from loading ai-examples extension which I haven't touched. Please give guidance on how to proceed.

Yes. We could fix the CI error.

**Comment by jaredwelsh** [comment]
Changes were made @Juliehzl

**Review by Juliehzl — APPROVED** [review]

**Pull request merged**